File: ontrack-service/src/test/java/net/nemerosa/ontrack/service/StandardBuildFilterProviderTest.java
package net.nemerosa.ontrack.service;

import net.nemerosa.ontrack.json.JsonUtils;
import net.nemerosa.ontrack.model.structure.PropertyService;
import net.nemerosa.ontrack.model.structure.StandardBuildFilterData;
import net.nemerosa.ontrack.model.structure.StructureService;
import net.nemerosa.ontrack.model.structure.ValidationRunStatusService;
import net.nemerosa.ontrack.repository.CoreBuildFilterRepository;
import org.junit.Before;
import org.junit.Test;

import java.util.Optional;

import static org.junit.Assert.*;
import static org.mockito.Mockito.mock;

public class StandardBuildFilterProviderTest {

    private StandardBuildFilterProvider provider;

    @Before
    public void before() {
        StructureService structureService = mock(StructureService.class);
        ValidationRunStatusService validationRunStatusService = mock(ValidationRunStatusService.class);
        PropertyService propertyService = mock(PropertyService.class);
        CoreBuildFilterRepository coreBuildFilterRepository = mock(CoreBuildFilterRepository.class);
        provider = new StandardBuildFilterProvider(
                structureService,
                validationRunStatusService,
                propertyService,
                coreBuildFilterRepository);
    }

    @Test
    public void parse_count_only() {
        Optional<StandardBuildFilterData> data = provider.parse(JsonUtils.object().with("count", 5).end());
        assertTrue(data.isPresent());
        assertEquals(5, data.get().getCount());
        assertNull(data.get().getWithPromotionLevel());
    }

    @Test
    public void parse_with_promotion_level_null() {
        Optional<StandardBuildFilterData> data = provider.parse(JsonUtils.object()
                .with("count", 5)
                .with("withPromotionLevel", (String) null)
                .end());
        assertTrue(data.isPresent());
        assertEquals(5, data.get().getCount());
        assertNull(data.get().getWithPromotionLevel());
    }

    @Test
    public void parse_with_after_date_null() {
        Optional<StandardBuildFilterData> data = provider.parse(JsonUtils.object()
                .with("count", 5)
                .withNull("afterDate")
                .end());
        assertTrue(data.isPresent());
        assertEquals(5, data.get().getCount());
        assertNull(data.get().getAfterDate());
    }

}

File: ontrack-service/src/test/java/net/nemerosa/ontrack/service/StandardBuildFilterProviderTest.kt
package net.nemerosa.ontrack.service

import net.nemerosa.ontrack.json.asJson
import net.nemerosa.ontrack.model.structure.PropertyService
import net.nemerosa.ontrack.model.structure.StructureService
import net.nemerosa.ontrack.model.structure.ValidationRunStatusService
import net.nemerosa.ontrack.repository.CoreBuildFilterRepository
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

class StandardBuildFilterProviderTest {

    private lateinit var provider: StandardBuildFilterProvider

    @BeforeEach
    fun before() {
        val structureService = mock(StructureService::class.java)
        val validationRunStatusService = mock(ValidationRunStatusService::class.java)
        val propertyService = mock(PropertyService::class.java)
        val coreBuildFilterRepository = mock(CoreBuildFilterRepository::class.java)
        provider = StandardBuildFilterProvider(
            structureService,
            validationRunStatusService,
            propertyService,
            coreBuildFilterRepository
        )
    }

    @Test
    fun `Parsing default`() {
        val data = provider.parse(
                emptyMap<String,String>().asJson()
        )
        assertNotNull(data) {
            assertEquals(10, it.count)
        }
    }

    @Test
    fun `Parsing without count`() {
        val data = provider.parse(
                mapOf(
                        "withPromotionLevel" to "IRON",
                ).asJson()
        )
        assertNotNull(data) {
            assertEquals(10, it.count)
            assertEquals("IRON", it.withPromotionLevel)
        }
    }
    @Test
    fun parse_count_only() {
        val data = provider.parse(
            mapOf("count" to 5).asJson()
        )
        assertNotNull(data) {
            assertEquals(5, it.count)
            assertNull(it.withPromotionLevel)
        }
    }

    @Test
    fun parse_with_promotion_level_null() {
        val data = provider.parse(
            mapOf(
                "count" to 5,
                "withPromotionLevel" to null,
            ).asJson()
        )
        assertNotNull(data) {
            assertEquals(5, it.count)
            assertNull(it.withPromotionLevel)
        }
    }

    @Test
    fun parse_with_after_date_null() {
        val data = provider.parse(
            mapOf(
                "count" to 5,
                "afterDate" to null,
            ).asJson()
        )
        assertNotNull(data) {
            assertEquals(5, it.count)
            assertNull(it.afterDate)
        }
    }

}